### PR TITLE
Chore: replace spectron with playwright

### PR DESCRIPTION
**Pull request by jackkav**
Related: https://github.com/electron-userland/spectron/issues/1045

The primary goal of this PR is to unblock our electron upgrade PR #4232 which is blocked by spectron flaking out on windows and blocking future upgrades due to its deprecation.

Improvements
- less flake in the playwright attaching to window mechanism
- runs twice as fast
- fails fast when missing build artifact

Scoped out
- screenshot/tracing on fail
- test writing guidance
- retry logic
- videos 

Notes:
- windows doesn't start properly so theres a forced reload in the test setup, this should be investigated separately
- locking to single instance doesn't work in playwright, this should be investigated separately
- ~~users who may have been using the undocumented env var INSOMNIA_DATA_PATH may get logged out because of the setPath fallback used for controlling migration context.~~ broke this out into a separate issue

Considered in this scope
- [x] docs, designed initially to be operated as spectron was
- [x] screenshot/video on fail, going to delay into a PR that adds the more powerful tracing support
- [x] locking issue, going to delay into an investigation
- [x] test setup fixtures, going to delay into the tracing PR above
- [x] better insomnia request collection fixture, not required
- [x] decide on playwright config, defaults are okay for now
- [x] decide if this PR should include code re-use abstraction, not yet

Closes INS-1300
Closes INS-1147

### CODE REVIEW

File: packages/insomnia-smoke-test/package.json [hunk 1]
```diff
@@ -19,16 +19,17 @@
     "clean": "tsc --build tsconfig.build.json --clean",
     "postclean": "rimraf dist",
     "build": "tsc --build tsconfig.build.json",
-    "spectron:build": "cross-env BUNDLE=build xvfb-maybe jest --detectOpenHandles --testPathPattern core",
-    "spectron:package": "cross-env BUNDLE=package xvfb-maybe jest --detectOpenHandles --testPathPattern core",
+    "playwright:build": "xvfb-maybe cross-env BUNDLE=build PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 npx playwright@1.17.1 test",
+    "playwright:package": "xvfb-maybe cross-env BUNDLE=package PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 npx playwright@1.17.1 test",
```
Comment (dimitropoulos) [inline packages/insomnia-smoke-test/package.json]: just curious what the reason was for flipping where `xvfb-maybe` is ordered in

File: packages/insomnia-smoke-test/specs/app.spec.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+type Playwright = PlaywrightWorkerArgs['playwright'];
+
+const loadFixture = async (fixturePath: string) => {
+  const buffer = await fs.promises.readFile(path.join(__dirname, '..', 'fixtures', fixturePath));
+  return buffer.toString('utf-8');
+};
+
+const insomniaBinaryPath = () => process.platform === 'win32' ? path.join('win-unpacked', 'Insomnia.exe')
+  : process.platform === 'darwin' ? path.join('mac', 'Insomnia.app', 'Contents', 'MacOS', 'Insomnia')
+    : path.join('linux-unpacked', 'insomnia');
```
Comment (dimitropoulos) [inline packages/insomnia-smoke-test/specs/app.spec.ts]: could we possibly un-nested-ternary this?

File: packages/insomnia-smoke-test/package.json (hunk 1 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/package.json]: The additional env var

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 2 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: Done-zo

File: packages/insomnia-app/app/main.development.ts [hunk 3]
```diff
@@ -37,7 +37,7 @@ log.info(`Running version ${getAppVersion()}`);
 if (!isDevelopment()) {
   const defaultPath = app.getPath('userData');
   const newPath = path.join(defaultPath, '../', appConfig.userDataFolder);
-  app.setPath('userData', newPath);
+  app.setPath('userData', process.env.INSOMNIA_DATA_PATH ?? newPath);
```
Comment (develohpanda) [inline packages/insomnia-app/app/main.development.ts:40]: Why is this needed? We have https://github.com/Kong/insomnia/blob/8918b1511383ecf2904d564ff50beb3228855ea8/packages/insomnia-app/app/common/electron-helpers.ts#L26-L29. Super wary of making data path changes.

File: packages/insomnia-smoke-test/README.md [hunk 4]
```diff
@@ -54,14 +54,14 @@ npm run serve
...
+npm run playwright:build         # Insomnia build tests
 
-npm run spectron:package       # Insomnia package tests
+npm run playwright:package       # Insomnia package tests
 ```
 
 This will allow you to write and monitor the server separately from each test, speeding up the development cycle.
 
-You may also need to run a test multiple times. You can focus a particular test or test suite using [Jest globals](https://jestjs.io/docs/en/api#testonlyname-fn-timeout) such as `it.only` and `describe.skip`, or their aliases `fit` and `xdescribe`, etc.
+If the playwright test runner doesn't start the following command will show the stack on launch `DEBUG=pw:browser`
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/README.md]: "The following command" or "the following environment variable?

File: packages/insomnia-smoke-test/README.md [hunk 5]
```diff
@@ -72,189 +72,3 @@ Individual tests will automatically run against a clean Insomnia data directory
...
-
-```ts
-describe.only.each(new Array(20).fill(1))('iteration %#', _ => {
-  it('your test name', () => {
-    //...
-  });
-});
-```
-
-When raising a PR, paste a screenshot of the test results showing at least 20 successful iterations.
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/README.md]: Please reinstate these lines, as they are not specific to Spectron

File: packages/insomnia-smoke-test/package.json [hunk 6]
```diff
@@ -19,16 +19,17 @@
     "clean": "tsc --build tsconfig.build.json --clean",
     "postclean": "rimraf dist",
     "build": "tsc --build tsconfig.build.json",
-    "spectron:build": "cross-env BUNDLE=build xvfb-maybe jest --detectOpenHandles --testPathPattern core",
-    "spectron:package": "cross-env BUNDLE=package xvfb-maybe jest --detectOpenHandles --testPathPattern core",
+    "playwright:build": "xvfb-maybe cross-env BUNDLE=build PLAYWRIGHT_SKIP_BROWSER_DOWNLOAD=1 npx playwright@1.17.1 test",
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/package.json]: Why use npx instead of installing playwright as a proper dependency? If it's a dependency, it will be checked by dependabot, for example

File: packages/insomnia-smoke-test/playwright.config.ts [hunk 7]
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable filenames/match-exported */
+import { PlaywrightTestConfig } from '@playwright/test';
+const config: PlaywrightTestConfig = {
+  webServer: {
+    command: 'npm run serve',
+    port: 4010,
+    timeout: 120 * 1000,
+    reuseExistingServer: !process.env.CI,
+  },
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/playwright.config.ts:9]: This is very cool!

File: packages/insomnia-smoke-test/playwright.config.ts [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+  timeout: 20 * 1000, // @TODO: shorten this after we fix the 5 second app start delay
+  forbidOnly: !!process.env.CI,
+  outputDir: 'screenshots',
+  testDir: 'specs',
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/playwright.config.ts]: Please call the "test" terminology instead of "spec" to a) differentiate between openapi specs, and b) follow the pattern this repository. `tests/foo.test.ts` or `__tests__/foo.test.ts`.

File: packages/insomnia-smoke-test/playwright/paths.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+export const mainPath = path.join('build', 'main.min.js');
+export const cwd = path.resolve(__dirname, '..', '..', 'insomnia-app');
+
+const hasMainBeenBuilt = fs.existsSync(path.resolve(cwd, mainPath));
+const hasBinaryBeenBuilt = fs.existsSync(path.resolve(cwd, insomniaBinary));
+
+// NOTE: guard against missing build artifacts
+if (!isPackageTest && !hasMainBeenBuilt) {
+  console.error(`ERROR: ${mainPath} not found at ${path.resolve(cwd, mainPath)}
+"npm run app-build:smoke"`);
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/playwright/paths.ts]:
```suggestion
  console.error(`ERROR: ${mainPath} not found at ${path.resolve(cwd, mainPath)}. Have you run "npm run app-build:smoke"?`);
```

File: packages/insomnia-smoke-test/playwright/paths.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  exit(1);
+}
+if (isPackageTest && !hasBinaryBeenBuilt) {
+  console.error(`ERROR: ${insomniaBinary} not found at ${path.resolve(cwd, insomniaBinary)} 
+"npm run app-package:smoke"`);
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/playwright/paths.ts]:
```suggestion
  console.error(`ERROR: ${insomniaBinary} not found at ${path.resolve(cwd, insomniaBinary)}. Have you run "npm run app-package:smoke"?`);
```

File: packages/insomnia-smoke-test/specs/app.spec.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
+import { PlaywrightWorkerArgs, test } from '@playwright/test';
+
+import { cwd, DESIGNER_DATA_PATH, executablePath, loadFixture, mainPath, randomDataPath } from '../playwright/paths';
+type Playwright = PlaywrightWorkerArgs['playwright'];
+const isPackageTest = process.env.BUNDLE === 'package';
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/specs/app.spec.ts]: This line also exists inside `paths.ts`, please consolidate

File: packages/insomnia-smoke-test/specs/app.spec.ts [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+console.log(`Using current working directory at ${cwd}`);
+console.log(`Using executablePath at ${executablePath}`);
+console.log(`Using mainPath at ${mainPath}`);
+
+const newPage = async ({ playwright, options }: ({ playwright: Playwright; options: {} })) => {
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/specs/app.spec.ts]: Given `options` is actually being used further down, please add a type definition instead of  `{}`

File: packages/insomnia-smoke-test/specs/app.spec.ts [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+  await electronApp.waitForEvent('window');
+  const page = await electronApp.firstWindow();
+  // page.on('console', console.log); a bit noisy but helpful for debugging
+  // @TODO: Investigate why the app doesn't start without a reload with playwright in windows
+  if (process.platform === 'win32') await page.reload();
+  return { electronApp, page };
+};
+
+test('can send requests', async ({ playwright }) => {
+  const options = { INSOMNIA_DATA_PATH: randomDataPath };
```
Comment (develohpanda) [inline packages/insomnia-smoke-test/specs/app.spec.ts]: It looks to me like `randomDataPath` will not change between tests. So, if I have 3 unique tests in this suite, they will all use the same data path. If this is correct, is this behaviour intentional, and is it appropriate? Why/why not?

**Review by develohpanda — CHANGES_REQUESTED** [review]
> less flake in the playwright attaching to window mechanism
> runs twice as fast

These two things are huge, well done on working through it 👏🏽 

I know you've chosen to descope writing utilities/page objects from this PR. In theory that's fine, it's a work in progress, but we're deleting a _lot_ of valuable code and if we slightly alter the approach we can take a step up from Spectron instead of start from scratch. Note that what we are changing is the test framework, and from what I've seen in this PR that doesn't need to involve writing things from scratch.

The existing page object approach of writing an e2e test is agnostic of the framework being used. If you look at the overall functions - click a request, click send, click a tab, open a dropdown - and how they are parameterized, they describe user actions, and the implementation describes how the test framework should handle those user actions.

We've changed our test framework from Spectron to Playwright. Updating the page object implementations to work with Playwright's API will allow us to keep the existing tests (which are much more readable than the what I'm reviewing in this PR) and continue from there.

I have a strong opinion that it is not necessary to delete all of the existing tests and page objects and start again. This is feedback about the overall approach, please don't take it as a request to increase the scope of this PR. How can we help ourselves build on top of what's already written, instead of writing from scratch?

(The changes requested relate to some of the comments below.)

File: packages/insomnia-smoke-test/package.json (hunk 6 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/package.json]: I was just applying a convention from cypress, to avoid dev deps that might contain binaries like chrome, but since @playwright/test contains playwright, and browser is disabled there is no need to use the convention, thanks for the push.

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 13 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: yep this was a mistake, thanks for the catch

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 12 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: fair

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 11 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: this function isn't related to paths I'll just remove it.

File: packages/insomnia-smoke-test/README.md (hunk 4 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/README.md]: thanks docs are hard

File: packages/insomnia-app/app/main.development.ts (hunk 3 above)
Comment (jackkav) [reply packages/insomnia-app/app/main.development.ts:40]: I tried it without this and it seems good, thanks for the push

File: packages/insomnia-smoke-test/README.md (hunk 5 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/README.md]: ah the inso CLI doc delete was a mistake, thanks for the catch.
I don't agree that this 20 times process is desirable. If anything it warns contributors away from contributing smoke tests, which playwright aims to encourage.

File: packages/insomnia-smoke-test/playwright.config.ts (hunk 7 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/playwright.config.ts:9]: Sure is, haven't figured out how to show the logs for it yet though. Haven't needed to yet.

File: packages/insomnia-smoke-test/playwright.config.ts (hunk 8 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/playwright.config.ts]: sure thing

**Comment by jackkav** [comment]
Thank you for your thoughts and concerns I hear you and hope to address them to your satisfaction.

> I know you've chosen to descope writing utilities/page objects from this PR. In theory that's fine, it's a work in progress, but we're deleting a lot of valuable code and if we slightly alter the approach we can take a step up from Spectron instead of start from scratch. Note that what we are changing is the test framework, and from what I've seen in this PR that doesn't need to involve writing things from scratch.

Playwright and Spectron are not a 1:1 API match, there are new playwright specific conventions to learn and adapt, and it is unwise to learn and adapt to them as one contributor in a single PR.

> The existing page object approach of writing an e2e test is agnostic of the framework being used. If you look at the overall functions - click a request, click send, click a tab, open a dropdown - and how they are parameterized, they describe user actions, and the implementation describes how the test framework should handle those user actions.

The reason this code uses playwright recordings for its runs are to separate any of my personal opinions about test structuring from the replacement effort, while creating an opportunity to revisit existing conventions, I agree the modules were very useful in making readable spectron tests possible, and we will certainly want to pull in some of the ideas with playwright. However I strongly believe the introduction of abstractions over test structure should happen carefully as these tests have a huge impact on the contribution experience, so we have a duty to keep them as lean and fast as possible to maximise return on investment.

> We've changed our test framework from Spectron to Playwright. Updating the page object implementations to work with Playwright's API will allow us to keep the existing tests (which are much more readable than the what I'm reviewing in this PR) and continue from there.

Readability is subjective and not something I aim to improve in this PR, only to provide us with a starting point to build confidence in our e2e testing.

> I have a strong opinion that it is not necessary to delete all of the existing tests and page objects and start again. This is feedback about the overall approach, please don't take it as a request to increase the scope of this PR. How can we help ourselves build on top of what's already written, instead of writing from scratch?

We can build on top by writing new tests ourselves on top of this minimal safety net, to show the value of required abstractions.

My goal is to eliminate barriers to contribution, to me this means making it intuitive, easy to change and with as little indirection as possible.

File: packages/insomnia-smoke-test/specs/app.spec.ts [hunk 14]
```diff
@@ -39,7 +39,6 @@ test('can send requests', async ({ playwright }) => {
   await electronApp.evaluate(async ({ clipboard }, text) => clipboard.writeText(text), text);
 
   await page.click('button:has-text("Clipboard")');
-  await page.click('text=Smoke tests');
```
Comment (filfreire) [inline packages/insomnia-smoke-test/specs/app.spec.ts]: @jackkav by removing this the tests seem to be failing now, since playwright can't open the collection to try and issue the requests.
Screenshot 2021-12-16 at 14 29 45

File: .github/workflows/test.yml [hunk 15]
```diff
@@ -61,6 +61,12 @@ jobs:
       - name: Run tests
         run: npm test
 
+      - name: Build app for smoke tests
+        run: npm run app-build:smoke
```
Comment (filfreire) [inline .github/workflows/test.yml:65]: To pick on something @dimitropoulos had mentioned a while back, I think we could just try and also change the commands both here and in the `package.json` to something that is more readable and simpler.
Related to this ticket: https://linear.app/insomnia/issue/INS-1276/improve-npm-script-experience

Perhaps reducing all our smoke commands to just 2, something like `npm run smoke:build` and `npm run smoke:test`? (Not sure about the naming though)

File: packages/insomnia-smoke-test/playwright.config.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+/* eslint-disable filenames/match-exported */
+import { PlaywrightTestConfig } from '@playwright/test';
+const config: PlaywrightTestConfig = {
+  webServer: {
```
Comment (filfreire) [inline packages/insomnia-smoke-test/playwright.config.ts:4]: Same opinion as @develohpanda on this, this is awesome. There's probably a way for us to mimic this and add one for gRPC, and later on websockets , etc. Really cool stuff 👍

File: packages/insomnia-smoke-test/tests/app.test.ts [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+  });
+  await electronApp.waitForEvent('window');
+  const page = await electronApp.firstWindow();
+  // @TODO: Investigate why the app doesn't start without a reload with playwright in windows
+  if (process.platform === 'win32') await page.reload();
+  return { electronApp, page };
+};
+
+test('can send requests', async ({ playwright }) => {
+  const options = { INSOMNIA_DATA_PATH: randomDataPath() };
```
Comment (filfreire) [inline packages/insomnia-smoke-test/tests/app.test.ts:31]: This is giving me an idea - let's say we get a bug that is only reproducible with someone's backup of a data folder, we could make it so folks can point to a data path for the test they want to run - which we would fetch locally or maybe download from S3 bucket where we store weird data folder backups.

File: packages/insomnia-smoke-test/tests/app.test.ts [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  const { page, electronApp } = await newPage({ playwright, options });
+  await page.click('text=Don\'t share usage analytics');
+  await page.click('text=Create');
+
+  const text = await loadFixture('smoke-test-collection.yaml');
+  await electronApp.evaluate(async ({ clipboard }, text) => clipboard.writeText(text), text);
+
+  await page.click('button:has-text("Clipboard")');
+  await page.click('button:has-text("GETsend JSON request")');
```
Comment (filfreire) [inline packages/insomnia-smoke-test/tests/app.test.ts:41]: I'm not sure if my other comment will be visible, but here we're missing the step before this one where we click to open the collection after importing.

**Review by filfreire — APPROVED** [review]
Marking this as approved, seems to work ok, just some minor comments:
- Tests were passing earlier now are failing, I suspect due to a missing click on the `can send requests` scenario
- Left some notes of stuff we might want to look at in the future - like spinning up gRPC and Websocket test servers like you did for http, and also we might want to explore tests where we purposely import insomnia data folder that are in a weird state.

File: packages/insomnia-smoke-test/README.md [hunk 19]
```diff
@@ -2,17 +2,16 @@
...
 
-| Folder | Purpose |
-| - | - |
-| `/cli` | tests for inso |
-| `/core` | tests for Insomnia |
-| `/server` | Express server used by the tests |
+| Folder      | Purpose                           |
+| ----------- | --------------------------------- |
+| `/cli`      | tests for inso                    |
+| `/specs`    | tests for Insomnia                |
```
Comment (DMarby) [inline packages/insomnia-smoke-test/README.md]:
```suggestion
| `/tests`    | tests for Insomnia                |
```

**Review by DMarby — APPROVED** [review]
Awesome 👍 

Found two remaining references to Spectron, not sure if we have the same issue with the `tsconfig` with Playwright as we did for Spectron?
https://github.com/Kong/insomnia/blob/b2c94ebbdbcc72750d1cb415f058d49a57ca5676/packages/insomnia-smoke-test/tsconfig.build.json#L9
https://github.com/Kong/insomnia/blob/b2c94ebbdbcc72750d1cb415f058d49a57ca5676/DEVELOPMENT.md?plain=1#L60

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 14 above)
Comment (DMarby) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: Can confirm this locally as well, adding this line back in makes the test pass for me

File: .github/workflows/test.yml (hunk 15 above)
Comment (jackkav) [reply .github/workflows/test.yml:65]: I like it!

File: packages/insomnia-smoke-test/tests/app.test.ts (hunk 17 above)
Comment (DMarby) [inline packages/insomnia-smoke-test/tests/app.test.ts:31]: I think we need to pass `DESIGNER_DATA_PATH` for this test as well (or preferably all tests by default) to a directory that does not exist, as otherwise the app will pick up the local designer directory if there is one and attempt to migrate (this breaks the tests for me locally since I have an old Insomnia Designer install present)

I believe we used to use the same approach for Spectron: https://github.com/Kong/insomnia/blob/b2c94ebbdbcc72750d1cb415f058d49a57ca5676/packages/insomnia-smoke-test/modules/application.ts#L19

File: .github/workflows/test.yml (hunk 15 above)
Comment (jackkav) [reply .github/workflows/test.yml:65]: It's in so many places in the code base and meant to exist alongside many other contexts so I'll scope it out for now, but worth a later PR to look for ways to clarify the wording of `app-build:smoke` and `test:smoke:build`

File: packages/insomnia-smoke-test/specs/app.spec.ts (hunk 14 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/specs/app.spec.ts]: I was seeing import from clipboard go straight to the send request view, so I assumed it had been changed in a recent PR after rebase. Not sure though.

File: packages/insomnia-smoke-test/tests/app.test.ts (hunk 17 above)
Comment (jackkav) [reply packages/insomnia-smoke-test/tests/app.test.ts:31]: done, thanks for the catch! I added a note too for if we remove migration

File: packages/insomnia-smoke-test/README.md (hunk 5 above)
Comment (develohpanda) [reply packages/insomnia-smoke-test/README.md]: Ah that's all good, we can drop the 20 times thing. We can bring it back if the need arises!

File: .github/workflows/test.yml (hunk 15 above)
Comment (develohpanda) [reply .github/workflows/test.yml:65]: I agree with this to try and increase clarity! There are a few variations hence the confusion:

* build the app
* build the app for smoke tests
* package the app
* package the app for smoke tests
* run smoke tests against build artifacts
* run smoke tests against the packaged binaries

Building the app is faster than packaging the app. But certain things (such as testing plugin installation) requires the test to be run against a packaged binary instead of build artifacts. Testing against the packaged app is also more realistic, so it's just weighing up the pros and cons.

Additionally, the first four bulletpoints can become _two_ bulletpoints if we can remove/address the requirement for a gitref in the build script: https://github.com/Kong/insomnia/blob/5f4c19da35e90ffa2f3ae27e9b309ee48f9d4e16/packages/insomnia-app/scripts/getBuildContext.ts#L38-L51

File: packages/insomnia-smoke-test/tests/app.test.ts (hunk 17 above)
Comment (develohpanda) [reply packages/insomnia-smoke-test/tests/app.test.ts:31]: Nice catch 🔥

**Review by develohpanda — APPROVED** [review]
Thanks for your thoughts as well! Readability _is_ subjective but readability helps with code reviews, and while the contribution experience is important, maintainability is also important when tests break and someone goes in with fresh eyes, and over time I hope we can reach a balance between all of these factors.

One action item: could you please make follow-up tasks to go to the existing smoke tests and migrate many of the assertions across?

For example, the PDF viewer assertion, or the XML filter assertion are both regression tests and we shouldn't lose those learnings. https://github.com/Kong/insomnia/blob/b2c94ebbdbcc72750d1cb415f058d49a57ca5676/packages/insomnia-smoke-test/core/app.test.ts#L81-L94.

Another example is the Basic Auth test. Instead of simply testing basic auth works through a header, it's purpose is to use the Auth tab within Insomnia to simulate a user typing into the fields, enabling and disabling auth and verifying the expected response.

tl;dr through this PR we lose a lot of coverage, and we should make an effort to bring across the existing tests once this has merged to reinstate that coverage, hence follow-up tasks.
